dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,408 @@
+    using Request = typename RPCContext::Request;
+    using AsyncReader = typename RPCContext::AsyncReader;
+    using Self = AsyncRequestHandler<RPCContext, enable_fine_grained_shuffle>;
+
+    AsyncRequestHandler(
+        std::vector<GRPCReceiveQueue<RecvMsgPtr>> & grpc_recv_queues_,
+        AsyncRequestHandlerWaitQueuePtr async_wait_rewrite_queue_,
+        const std::shared_ptr<RPCContext> & context,
+        Request && req,
+        const String & req_id,
+        std::atomic<Int64> * data_size_in_queue,
+        std::function<void(bool, const String &, const LoggerPtr &)> && close_conn_)
+        : cq(&(GRPCCompletionQueuePool::global_instance->pickQueue()))
+        , rpc_context(context)
+        , request(std::move(req))
+        , req_info(fmt::format("tunnel{}+{}", req.send_task_id, req.recv_task_id))
+        , has_data(false)
+        , retry_times(0)
+        , stage(AsyncRequestStage::NEED_INIT)
+        , read_packet_index(0)
+        , received_packet_index(0)
+        , finish_status(RPCContext::getStatusOK())
+        , log(Logger::get(req_id, req_info))
+        , channel_try_writer(grpc_recv_queues_, req_info, log, data_size_in_queue, ReceiverMode::Async)
+        , async_wait_rewrite_queue(std::move(async_wait_rewrite_queue_))
+        , kick_recv_tag(thisAsUnaryCallback())
+        , close_conn(std::move(close_conn_))
+        , is_close_conn_called(false)
+    {
+        FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::exception_when_construct_async_request_handler);
+        packets.resize(batch_packet_count);
+        for (auto & packet : packets)
+            packet = std::make_shared<TrackedMppDataPacket>(MPPDataPacketV0);
+        start();
+    }
+
+    // execute will be called by RPC framework so it should be as light as possible.
+    // Do not do anything after processXXX functions.
+    void execute(bool & ok) override
+    {
+        try
+        {
+            switch (stage)
+            {
+            case AsyncRequestStage::WAIT_RETRY:
+                start();
+                break;
+            case AsyncRequestStage::WAIT_MAKE_READER:
+                processWaitMakeReader(ok);
+                break;
+            case AsyncRequestStage::WAIT_BATCH_READ:
+                processWaitBatchRead(ok);
+                break;
+            case AsyncRequestStage::WAIT_REWRITE:
+                processWaitReWrite();
+                break;
+            case AsyncRequestStage::WAIT_FINISH:
+                processWaitFinish();
+                break;
+            default:
+                RUNTIME_ASSERT(false, "Unexpected stage {}", magic_enum::enum_name(stage));
+            }
+        }
+        catch (...)
+        {
+            LOG_ERROR(log, "Some errors happen in the AsyncRequestHandler.execute()");
+            closeConnection("Exception is thrown in AsyncRequestHandler");
+        }
+    }
+
+    void wait() override
+    {
+        std::unique_lock<std::mutex> ul(is_close_conn_called_mu);
```

Comment:
maybe use lock_guard?

Reply:
> maybe use lock_guard?

lock passed into condition_variable must be unique_lock